### PR TITLE
Support simple query pushdown

### PR DESCRIPTION
**Pull request by seeforschauer**
# Description

This is analysis PR for #2, we won't merge it.
Select queries of the form `{_id: <value>}`. 
PR scope is `{_id: <value>}` where `<value>` is `ObjectID`.

## Tigris

TODO: How will we ensure that Tigris compares values the same way as MongoDB?

## Not in scope

Other cases when `<value>` of types are not in scope:
* Strings
  * string comparison difference between MongoDB and PostgreSQL i.e.
  * case sensitive
  * zero values
  * collation
  * encoding checks
  * etc.
* Binary data
  * len
  * sub-type 
  * byte by byte 
* Numeric
* Dates and Timestamps
  * ordering 
* Non-existent Fields (NaN, Inf)
* Objects
* Arrays
  * arrays are restricted to be in `_id`

# Links
[The pushdown term](https://www.quora.com/What-do-we-mean-when-we-say-SQL-pushdown).
[MongoDB restrictions on `_id`](https://www.mongodb.com/docs/manual/reference/limits/).
[MongoDB Comparison/Sort Order](https://www.mongodb.com/docs/manual/reference/bson-type-comparison-order/).

### CODE REVIEW

**Comment by fcoury** [comment]
Please take a look at the work done on this PR as I think it overlaps a lot with the dynamic query generation needed for this feature:

https://github.com/FerretDB/FerretDB/pull/850

**Review by rumyantseva — COMMENTED** [review]
Personally, I'm completely unfamiliar with pushdown, and googling didn't help much. I managed to find some "aggregation push-down framework" but I don't really understand how it helps to solve the described problem.

Let's maybe describe what "pushdown" is?
Also, the problem definition says that it's all about memory usage. The solution part doesn't really clarify what will be in memory with the new approach, and why it is better.

File: adr/query_pushdown.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+* `{_id: <value>}`
+
+## The problem
+Our fetch methods read all documents before returning them to the user.
+That requires too much memory.
+
+## Solution
+
+Let's go step by step and first, implement a simple query pushdown for queries containing `{_id: <value>}`,
+i.e. add SQL condition `_id = $<placeholder>` in WHERE clause passed to the PostgreSQL backend.
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply adr/query_pushdown.md]: You can do this inside `mongosh`:
```
test_id> db.test.insertOne({"_id": 1.23})
{ acknowledged: true, insertedId: 1.23 }
test_id> db.test.insertOne({"_id": "123"})
{ acknowledged: true, insertedId: '123' }
test_id> db.test.insertOne({"_id": [1]})
{ acknowledged: true, insertedId: [ 1 ] }
test_id> db.test.insertOne({"_id": {"foo": "bar"}})
{ acknowledged: true, insertedId: { foo: 'bar' } }
test_id> db.test.find()
[
  { _id: 1.23 },
  { _id: '123' },
  { _id: [ 1 ] },
  { _id: { foo: 'bar' } }
]
```
We are supporting all these `_id` values in FerretDB.

**Comment by seeforschauer** [comment]
> Let's maybe describe what "pushdown" is? 

Here is the pushdown description:

> Let's go step by step and first, implement a simple query pushdown for queries containing `{_id: <value>}`,
> i.e. add SQL condition `_jsonb->`+ `p.Next()` + `=` in WHERE clause passed to the PostgreSQL backend.

Next question:
> Also, the problem definition says that it's all about memory usage.
> The solution part doesn't really clarify what will be in memory with the new approach, and why it is better.

Should I remove it? I think everyone understands that's not good to get all the data from the collection just to delete or update some fields...

File: adr/query_pushdown.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlekSi) [reply adr/query_pushdown.md]: > My impression of the task was: that there is no new functionality from the user perspective.

There is no new functionality from user perspective – we already support `_id` values that are documents, for example, and that should not change. The text should show how that would work

**Comment by AlekSi** [comment]
Good answer to what "pushdown" is: https://qr.ae/pv48xk

File: adr/query_pushdown.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+insert into test values('{"_id": 1}');
+insert into test values(`{"_id": "s"}`);
+insert into test values('{"_id": {"foo": "bar"} }');
+insert into test values('{"_id": {"$f":"NaN"} }');
+insert into test values('{"_id": {"$f":"-Infinity"} }');
+insert into test values('{"_id": null }');
+insert into test values('{"_id": [1] }');
+insert into test values('{"_id": [null] }');
+insert into test values('{"_id": { "_id": [null] } }');
+insert into test values('{"_id": { "_id": { "$f":"NaN"} } }');
```
[the remark this answers is not in the document]
Comment (seeforschauer) [reply adr/query_pushdown.md:65]: > Inserts are already implemented, there is no need to write about them again

* they must be checked that they are working for this task.
* I found this list a helpful example

File: adr/query_pushdown.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+```
+
+[PostgreSQL functions](https://www.postgresql.org/docs/14/functions-json.html)
+
+## Tigris
+
+Support tables where the primary key is only one field.
+
+* `if len(schema.PrimaryKey) > 1` fallback to fetch the entire table.
+* vlaue type if not `ObjectID` raise error.
```
Comment (rumyantseva) [inline adr/query_pushdown.md:25]: What do you mean by `ObjectID`? A specific type, a set of types? I don't see `ObjectID` mentioned in the article linked above.

File: adr/query_pushdown.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+Select queries of the form `{_id: <ObjectID>}` (value type is `ObjectID`).
+
+## Postgres
+
+If value type is not `ObjectID`, fallback to fetch entire table.
```
Comment (rumyantseva) [inline adr/query_pushdown.md:5]: What do you mean by `ObjectID`? A specific type, a set of types? I don't see `ObjectID` mentioned in the article linked above.

File: adr/query_pushdown.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+If value type is `ObjectID` raise error (`fjson` unmarshal).
```
Comment (rumyantseva) [inline adr/query_pushdown.md:6]: I don't understand that sentence. Does it mean "If value type is `ObjectID` then raise error"?

Comment (seeforschauer) [reply adr/query_pushdown.md:6]: > I don't understand that sentence. Does it mean "If value type is `ObjectID` then raise error"?

ooops, I meant if the value type is correct but the data in the value is corrupted, the raise error. (fixing)

File: adr/query_pushdown.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+Both work:
+
+```sql
+select * from test where (_jsonb->'_id')::jsonb->>'$o' = '507f1f77bcf86cd799439011';
```
Comment (rumyantseva) [inline adr/query_pushdown.md:11]: The example looks good to me, but I don't understand how it's connected to the two previous sentences. I think the part I miss is where we speak "FerretDB" language and where we speak "PotgreSQL" language.

**Review by w84thesun — COMMENTED** [review]
Overall looks good, the only question I have is about tests. Should we mention some? For example, we could add a unit test to check that we fetching the entire table if `_id` is not `ObjectID`.

**Comment by seeforschauer** [comment]
> Overall looks good, the only question I have is about tests. Should we mention some? 
> For example, we could add a unit test to check that we fetch the entire table if `_id` is not `ObjectID`.


Well, there definitely should be tests. How do you know that it was the query with `_id` only? i.e. we filter the data inside `fetch`  so if you add the test for the fetch, it won't check that that was the pushdown. What are your suggestions?

File: adr/query_pushdown.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
+Select queries of the form `{_id: <ObjectID>}` (value type is `ObjectID`).
+
+## Postgres
+
+If value type is not `ObjectID`, fallback to fetch entire table.
```
Comment (AlekSi) [inline adr/query_pushdown.md:5]: That requires more explanation: what fallback, how we will do it, etc. What will high-level changes in the source code be needed?

File: adr/query_pushdown.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+Proof of concept for a `{_id: <ObjectID>}` pushdown query, PostgreSQL:
+
+```sql
+select * from test where (_jsonb->'_id')::jsonb->>'$o' = '507f1f77bcf86cd799439011';
+```
+
+```sql
+select * from test where _jsonb->'_id' = '{"$o":"507f1f77bcf86cd799439011"}'::jsonb;
+```
```
Comment (AlekSi) [inline adr/query_pushdown.md:16]: Ok, so which one will we use and why?

File: adr/query_pushdown.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+[PostgreSQL functions](https://www.postgresql.org/docs/14/functions-json.html)
+
+## Tigris
+
+Support tables where the primary key is only one field.
+
+* if `len(schema.PrimaryKey) > 1` raise error.
+  * because user expects single one value in `_id`.
+  * to comply and be compatible with MongoDB.
+* if valaue type if not `ObjectID` raise error.
```
Comment (AlekSi) [inline adr/query_pushdown.md:27]: Why error and not a fallback like with PostgreSQL?

**Comment by seeforschauer** [comment]
Indexes tried: 
* `(_jsonb ->> '_id')::text`
* `((_jsonb ->> '_id' )::jsonb->> '$o')::text`
* `(_jsonb#>>'{_id,$o}')`

```sql
CREATE TABLE values (
	_jsonb jsonb NOT NULL DEFAULT '{}'
);

CREATE INDEX values_id_idx ON values (((_jsonb ->> '_id')::text));
insert into values values('{"_id": {"$o": "507f1f77bcf86cd799439011"} }');

---
CREATE INDEX values_id_idx ON values (((_jsonb ->> '_id')::text));
ferretdb=# EXPLAIN ANALYZE select * from values where (_jsonb->'_id')::jsonb->>'$o' = '507f1f77bcf86cd799439011';
                                             QUERY PLAN
----------------------------------------------------------------------------------------------------
 Seq Scan on "values"  (cost=0.00..33.80 rows=7 width=32) (actual time=0.031..0.033 rows=1 loops=1)
   Filter: (((_jsonb -> '_id'::text) ->> '$o'::text) = '507f1f77bcf86cd799439011'::text)
 Planning Time: 0.158 ms
 Execution Time: 0.849 ms
(4 rows)

ferretdb=# EXPLAIN ANALYZE select * from values where _jsonb->'_id' = '{"$o":"507f1f77bcf86cd799439011"}'::jsonb;
                                             QUERY PLAN
----------------------------------------------------------------------------------------------------
 Seq Scan on "values"  (cost=0.00..30.40 rows=7 width=32) (actual time=0.023..0.025 rows=1 loops=1)
   Filter: ((_jsonb -> '_id'::text) = '{"$o": "507f1f77bcf86cd799439011"}'::jsonb)
 Planning Time: 0.167 ms
 Execution Time: 0.053 ms
(4 rows)


---
CREATE INDEX values_id_idx ON values ( (((_jsonb ->> '_id' )::jsonb->> '$o')::text ));
ferretdb=# CREATE INDEX values_id_idx ON values ( (((_jsonb ->> '_id' )::jsonb->> '$o')::text ));
CREATE INDEX
ferretdb=# EXPLAIN ANALYZE select * from values where (_jsonb->'_id')::jsonb->>'$o' = '507f1f77bcf86cd799439011';
                                            QUERY PLAN
---------------------------------------------------------------------------------------------------
 Seq Scan on "values"  (cost=0.00..1.02 rows=1 width=32) (actual time=0.024..0.026 rows=1 loops=1)
   Filter: (((_jsonb -> '_id'::text) ->> '$o'::text) = '507f1f77bcf86cd799439011'::text)
 Planning Time: 0.511 ms
 Execution Time: 0.058 ms
(4 rows)

ferretdb=# EXPLAIN ANALYZE select * from values where _jsonb->'_id' = '{"$o":"507f1f77bcf86cd799439011"}'::jsonb;
                                            QUERY PLAN
---------------------------------------------------------------------------------------------------
 Seq Scan on "values"  (cost=0.00..1.01 rows=1 width=32) (actual time=0.029..0.032 rows=1 loops=1)
   Filter: ((_jsonb -> '_id'::text) = '{"$o": "507f1f77bcf86cd799439011"}'::jsonb)
 Planning Time: 0.147 ms
 Execution Time: 0.066 ms


---------------
ferretdb=# CREATE INDEX values_id_idx ON values(( _jsonb#>>'{_id,$o}'));
CREATE INDEX
ferretdb=# SELECT
    tablename,
    indexname,
    indexdef
FROM
    pg_indexes
WHERE
    schemaname = 'public'
ORDER BY
    tablename,
    indexname;
 tablename |   indexname   |                                           indexdef
-----------+---------------+-----------------------------------------------------------------------------------------------
 values    | values_id_idx | CREATE INDEX values_id_idx ON public."values" USING btree (((_jsonb #>> '{_id,$o}'::text[])))
(1 row)
ferretdb=# EXPLAIN ANALYZE select * from values where (_jsonb->'_id')::jsonb->>'$o' = '507f1f77bcf86cd799439011';
                                            QUERY PLAN
---------------------------------------------------------------------------------------------------
 Seq Scan on "values"  (cost=0.00..1.02 rows=1 width=32) (actual time=0.025..0.028 rows=1 loops=1)
   Filter: (((_jsonb -> '_id'::text) ->> '$o'::text) = '507f1f77bcf86cd799439011'::text)
 Planning Time: 0.120 ms
 Execution Time: 0.058 ms
(4 rows)

ferretdb=# EXPLAIN ANALYZE select * from values where (_jsonb->'_id')::jsonb->>'$o' = '507f1f77bcf86cd799439011';
                                            QUERY PLAN
---------------------------------------------------------------------------------------------------
 Seq Scan on "values"  (cost=0.00..1.02 rows=1 width=32) (actual time=0.025..0.028 rows=1 loops=1)
   Filter: (((_jsonb -> '_id'::text) ->> '$o'::text) = '507f1f77bcf86cd799439011'::text)
 Planning Time: 0.117 ms
 Execution Time: 0.057 ms
(4 rows)
```

**Comment by seeforschauer** [comment]
better use GIN https://www.postgresql.org/docs/current/datatype-json.html#JSON-INDEXING

**Comment by seeforschauer** [comment]
it's not applied anyway :| 
```sql
ferretdb=# CREATE INDEX values_id_idx ON values USING GIN ((_jsonb->'_id'));
ferretdb=# SELECT
    tablename,
    indexname,
    indexdef
FROM
    pg_indexes
WHERE
    schemaname = 'public'
ORDER BY
    tablename,
    indexname;
 tablename |   indexname   |                                      indexdef
-----------+---------------+-------------------------------------------------------------------------------------
 values    | values_id_idx | CREATE INDEX values_id_idx ON public."values" USING gin (((_jsonb -> '_id'::text)))

ferretdb=# EXPLAIN ANALYZE select * from values where _jsonb->'_id' @> '{"$o":"507f1f77bcf86cd799439011"}';
                                            QUERY PLAN                                             
---------------------------------------------------------------------------------------------------
 Seq Scan on "values"  (cost=0.00..1.01 rows=1 width=32) (actual time=0.026..0.029 rows=1 loops=1)
   Filter: ((_jsonb -> '_id'::text) @> '{"$o": "507f1f77bcf86cd799439011"}'::jsonb)
 Planning Time: 0.149 ms
 Execution Time: 0.060 ms
(4 rows)
```

**Comment by seeforschauer** [comment]
Well, 

* It's probably doesn't use the index, since for the small queries it's more effective to use seq scan.
  To see how it will look like on large tables I tried `set enable_seqscan=false`: and anyway it used seq scan:

```sql
ferretdb=# CREATE INDEX values_id_idx ON public."values" USING gin ((((_jsonb -> '_id'::text) -> '$o'::text)))
ferretdb-# ;
CREATE INDEX
ferretdb=# set enable_seqscan=false;
SET
ferretdb=# EXPLAIN ANALYZE select * from values where ((_jsonb->'_id'::text)->'$o')::text = '507f1f77bcf86cd799439011';
;                                                      QUERY PLAN                                                       
-----------------------------------------------------------------------------------------------------------------------
 Seq Scan on "values"  (cost=10000000000.00..10000000040.60 rows=7 width=32) (actual time=0.008..0.009 rows=0 loops=1)
   Filter: ((((_jsonb -> '_id'::text) -> '$o'::text))::text = '507f1f77bcf86cd799439011'::text)
 Planning Time: 0.721 ms
 JIT:
   Functions: 2
   Options: Inlining true, Optimization true, Expressions true, Deforming true
   Timing: Generation 0.310 ms, Inlining 0.000 ms, Optimization 0.000 ms, Emission 0.000 ms, Total 0.310 ms
 Execution Time: 770.581 ms
(8 rows)
```
  
* Type casts for the case are matching: 
```sql
CREATE INDEX values_id_idx ON public."values" USING gin ((((_jsonb -> '_id'::text) -> '$o'::text)));
EXPLAIN ANALYZE select * from values where ((_jsonb->'_id'::text)->'$o')::text = '507f1f77bcf86cd799439011';
```
* Maybe also to experiment with the planner settings.

File: adr/query_pushdown.md (hunk 8 above)
Comment (AlekSi) [reply adr/query_pushdown.md:39]: Ok, let's just use the first one just because it is first :)

**Review by w84thesun — DISMISSED** [review]
LGTM.

**Review by w84thesun — APPROVED** [review]

**Review by noisersup — APPROVED** [review]

**Review by rumyantseva — CHANGES_REQUESTED** [review]
The PR is not ready for review, so I reject it.

**Pull request closed without merging**